pkg/crawler/crawler.go

```diff
@@ -0,0 +1,245 @@
+package crawler
+
+import (
+	"context"
+	"encoding/xml"
+	"fmt"
+	"github.com/aquasecurity/trivy-java-db/metadata"
+	"github.com/aquasecurity/trivy-java-db/pkg/db"
+	"github.com/aquasecurity/trivy-java-db/pkg/types"
+	"io"
+	"log"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/hashicorp/go-retryablehttp"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/xerrors"
+	"k8s.io/utils/clock"
+)
+
+const mavenRepoURL = "https://repo.maven.apache.org/maven2/"
+
+type Crawler struct {
+	wg             sync.WaitGroup
+	urlCh          chan string
+	indexCh        chan *types.Index
+	saveIndexesWG  sync.WaitGroup
+	tickerDuration time.Duration
+	limit          *semaphore.Weighted
+	client         *retryablehttp.Client
+	clock          clock.Clock
+}
+
+type Option struct {
+	Limit int64
+}
+
+func NewCrawler(opt Option) Crawler {
+	client := retryablehttp.NewClient()
+	client.Logger = nil
+	return Crawler{
+		urlCh:          make(chan string, opt.Limit*10),
+		indexCh:        make(chan *types.Index, opt.Limit),
+		tickerDuration: 500 * time.Millisecond,
+		limit:          semaphore.NewWeighted(opt.Limit),
+		client:         client,
+		clock:          clock.RealClock{},
+	}
+}
+
+func (c *Crawler) Crawl(ctx context.Context) error {
+	errCh := make(chan error)
+	defer close(errCh)
+
+	// Add a root url
+	c.urlCh <- mavenRepoURL
+	c.wg.Add(1)
+
+	go func() {
+		c.wg.Wait()
+		close(c.indexCh)
+		close(c.urlCh)
+		c.saveIndexesWG.Wait()
+	}()
+
+	go func() { // function for saving indexes in the database
```

Comment:
Do we need to run it in goroutine? I think DB update is fast enough compared to HTTP.

Reply:
we need 2 parallel loops:
1) to parse all URLS(https://github.com/aquasecurity/trivy-java-db/pull/1/files#diff-526a7142f7b15a72eab215dcfa540f5b6d502951a2f9f3b34dd900013e82c675R92)
2) to save indexes to DB.

1 loop works in main gorutine
Therefore, I run saved indexes in separate goroutine.

Reply:
URL parsing is already ran in goroutine. Why do we need another goroutine?

Reply:
GORM with SQLite is slow. That is why we need to use separate goroutine.